### PR TITLE
perf(core): rank gridded EDDI and percentage of normal once per spatial block (#942)

### LINKED ISSUE #942: VECTORIZE-893: Convert EDDI and percentage-of-normal to spatial apply_ufunc blocks
Part of the #893 performance epic. Follow-up from the spatial `apply_ufunc` conversion in #923.

## Problem

#923 adds a spatial kernel path to the `xarray_adapter` decorator (one `xr.apply_ufunc`
call per non-core block instead of one per grid cell) and enables it for SPI and SPEI.
`indices.eddi` and `indices.percentage_of_normal` share the same decorated generic
adapter path, so they still take the per-cell path from the #922 inventory:

- `xarray_adapter.py` (Dask branch and in-memory branch) — `vectorize=True`, one call per
  cell, 3306 calls for the 38x87 reference grid.
- `indices.eddi` ranks every year against its calendar period inside a per-period,
  per-year Python loop (`indices.py`, the `for period_index` / `for year_index` pair),
  which has no cell axis at all today.
- `indices.percentage_of_normal` divides the scaled sums by the per-calendar-step normals
  with explicit `(years, periods)` reshaping.

## Tasks

- [ ] Convert `indices.eddi` to a cell-axis ranking (e.g. `searchsorted`/comparison counts
      per calendar period over a `(years, *cells)` climatology), keeping the NOAA
      Tukey-plotting-position and leading-scale-pad semantics identical.
- [ ] Convert `indices.percentage_of_normal` to broadcast the per-calendar-step normals
      across cells instead of reshaping to `(years, periods)`.
- [ ] Enable the spatial kernel on the `additional` adapter call sites in
      `typed_public_api.py` (`_wrapped_eddi`, `_wrapped_percentage_of_normal`).
- [ ] Watch the memory ceiling of the ranking comparison: a `(years, calibration_years,
      cells)` boolean intermediate is O(years^2 x cells), so cells may need block-wise
      processing (document the block size, or use a sorted/`searchsorted` formulation).

## Acceptance criteria

- [ ] Gridded EDDI and percentage-of-normal run with no Python-level loop over grid cells
      (extend the no-loop guard added in #923).
- [ ] Results match the single-series path within the tolerance asserted in #923's
      equivalence tests.
- [ ] `uv run ruff check src/ tests/`, `uv run ruff format --check src/ tests/`,
      `uv run mypy src/`, and `uv run pytest` pass.

## Related

- #923 — the spatial `apply_ufunc` conversion and its no-loop guard.
- #922 / PR #936 — the inventory that lists these sites.
- #929 (baselines), #930 (equivalence tests).

### PR DESCRIPTION
**Pull request by monocongo**
## Summary

Converts the two remaining generic-adapter per-cell paths from the #922 inventory: `indices.eddi` and `indices.percentage_of_normal` now accept the same time-major `(time, *cells)` block that `spi`/`spei` and the PET adapters take, and `typed_public_api` wires both adapters with `spatial_kernel=True`, so a gridded EDDI or percentage-of-normal run costs one kernel call per non-core block instead of one call per grid cell.

Closes #942.

- **EDDI** replaces the `for period_index` / `for year_index` Python loop with one pass per calendar period. Each cell's rank is the count of its calibration-climatology values below the current value, so the count is a comparison summed over climatology rows; the rows are walked in chunks sized by a 4 MB element budget so the `(climatology years, years, *cells)` intermediate does not grow with the calibration length. The NOAA zero-based rank, the Tukey plotting position, the leading scale-pad offset, and the `[-3.09, 3.09]` clip are unchanged, and the NOAA reference suite (`tests/test_noaa_eddi_reference.py`) still passes.
- **Percentage of normal** averages each cell's calendar-period normals over the block and divides whole periods by them, with the trailing partial period using the normals of the steps it covers.
- **Shape contract**: both indices keep their legacy `DataShapeError` for a 3-D or higher input, so any block has to be declared with `spatial_time_major=True` (which the adapter always sets). This is deliberate: these indices' dimension errors are pinned to `DataShapeError` by existing tests (`tests/test_eddi.py`, `tests/test_indices.py`), unlike the fitting-based kernels' `ValueError`. A declared block is accepted, an undeclared one raises with the declaration named in the message.
- Two guards that measured `values.size` now measure the time axis: the "calibration period fits within the data" check (with its message) and the whole/remainder period split, both of which a block's cell count would otherwise inflate. The remainder split was entered on every block even with no partial period.

## Scope decisions

- No new public keyword on `typed_public_api` entry points: `spatial_time_major` stays internal to the NumPy API, exactly as for SPI/SPEI/PET.
- The rank comparison is the kernel's memory ceiling: one climatology chunk of the `(climatology years, years, *cells)` boolean at a time, ~4 MB. A sorted-`searchsorted` formulation per cell was considered and dropped — NumPy's `searchsorted` takes 1-D `a` only, so it would need a per-cell loop or a merge-based rank that is harder to verify against the NOAA oracle. The chunk budget is the simpler exact formulation.
- EDDI's period loop remains a loop over calendar periods (12 monthly, 366 daily), not over grid cells or years.

## Verification

- `uv run ruff check src/ tests/` — clean
- `uv run ruff format --check src/ tests/` — clean
- `uv run mypy src/` — clean
- `uv run pytest` — 2095 passed, 423 deselected
- `uv run pytest tests/test_release_integrity.py tests/test_review_scripts.py` — passed
- `uv run scripts/generate_llms_txt.py` — bundles regenerated for `docs/xarray_compatibility.md`

`tests/test_spatial_kernel.py` (55 tests, 15 new) pins:

- the no-loop guard for both indices through the wired `typed_public_api` entry points: one ranking pass / one preparation-seam call for a 3 x 2 grid;
- equivalence with the single-series NumPy path and with the per-cell adapter, including NaN cells, a partial final period, daily grids (366-day calendar plan), and Dask blocks vs in-memory;
- the all-missing block short-circuit (asserted with warnings as errors, since the averaging path would emit "Mean of empty slice");
- the undeclared-block `DataShapeError` naming `spatial_time_major`.

The spatial-contract tripwire (`test_kernel_without_the_spatial_contract_fails_loudly`) now uses a locally defined kernel without the declaration instead of `indices.percentage_of_normal`, which no longer lacks it.

## Summary by Sourcery

Run EDDI and percentage-of-normal calculations over declared spatial blocks while preserving legacy NumPy behavior and bounded memory use.

New Features:
- Enable gridded EDDI and percentage-of-normal calculations over time-major spatial blocks, reducing adapter work from per-cell calls to one call per spatial block.
- Add explicit spatial-block declarations for the two NumPy index APIs while preserving legacy shape validation for undeclared multidimensional inputs.

Bug Fixes:
- Prevent masked values from leaking into scaled calculations and ensure missing spatial cells do not produce averaging warnings or invalid results.
- Correct calibration-period and whole/remainder-period checks to use the block time axis rather than total elements.

Enhancements:
- Vectorize EDDI ranking across cells with bounded memory while preserving NOAA ranking, plotting-position, clipping, and calendar behavior.
- Compute percentage-of-normal calibration normals and partial trailing periods independently for each spatial cell.
- Centralize calculation-completion logging and short-circuit entirely missing spatial blocks.

Documentation:
- Update spatial-kernel documentation, ADR guidance, benchmark notes, and xarray compatibility documentation for EDDI and percentage of normal.

Tests:
- Add comprehensive spatial-kernel coverage for equivalence, partial periods, daily calendars, missing data, masked inputs, Dask execution, chunked ranking, shape contracts, and adapter call counts.

Chores:
- Regenerate the LLM documentation bundles.

## Summary by CodeRabbit

- **New Features**
  - EDDI and percentage of normal now support spatial, time-major blocks and process all cells in one pass.
  - Added spatial-kernel support for these indices in xarray and Dask workflows.
  - Added an option to explicitly interpret 3-D inputs as time-major spatial blocks.

- **Bug Fixes**
  - Improved handling of missing values, masked blocks, and empty spatial inputs.
  - Undeclared 3-D inputs now raise a consistent shape error.

- **Documentation**
  - Updated spatial-kernel compatibility guidance, execution notes, and benchmark information.

### CODE REVIEW

File: src/climate_indices/indices.py [hunk 1]
```diff
@@ -347,55 +388,57 @@ def eddi(
...
+        # multiply the block by the number of calibration years. Missing climatology
+        # values never compare below a value, so they stay out of the count.
+        climatology = pet_values[calibration_start_year_index : calibration_end_year_index + 1]
+        climatology_valid_counts = np.count_nonzero(~np.isnan(climatology), axis=0)
+        leading_pads_count = np.count_nonzero(
+            leading_scale_pads[calibration_start_year_index : calibration_end_year_index + 1],
+            axis=0,
+        )
+        cells_per_time_step = int(np.prod(pet_values.shape[2:], dtype=np.int64)) or 1
+        rows_per_chunk = max(1, _EDDI_RANK_COMPARISON_ELEMENT_BUDGET // (num_years * cells_per_time_step))
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/indices.py]: Fixed in 3b040d59: the chunk size was still computed from `num_years * cells_per_time_step`, so a single climatology row at the minimum chunk size of 1 could itself exceed the 4M-element budget for a wide enough spatial block. Chunking is now done across cells instead of climatology rows, since the time/calibration dimensions are bounded by the input series length while spatial width is caller-controlled. Verified with a 350x350 grid (122,500 cells, 40-year block): each comparison chunk now stays at exactly 4,000,000 elements regardless of grid width.

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** [PR #977 — perf(core): rank gridded EDDI and percentage of normal once per spatial block (#942)](https://github.com/monocongo/climate_indices/pull/977)
- **Lanes run:** 5 independent read-only lanes — correctness/edge cases, dataflow/layout contracts, performance/memory, tests/false-green, maintainability/docs. Two returned on the first provider; the other three were retried and completed on a second authenticated provider after that provider hit its usage limit (one lane also needed a different model pin, since its configured model is unsupported on a ChatGPT account).
- **Validation:** `ruff check`, `ruff format --check`, `mypy src/` clean; `pytest` 2100 passed / 423 deselected; `pytest tests/test_release_integrity.py tests/test_review_scripts.py` 45 passed; `scripts/generate_llms_txt.py` run after each doc edit. Correctness lane additionally differential-tested the new kernels against a reconstruction of the pre-change per-period/per-year loops (exact to 1e-10 across scales, calendars, calibration offsets, ties, NaNs, partial periods, float32/int64/masked inputs).
- **Fix commits:** [9e25a76b — perf(core): hold EDDI's probability tail one calendar period at a time](https://github.com/monocongo/climate_indices/commit/9e25a76b), [3b040d59 — fix(core): bound EDDI's rank comparison across cells, not climatology rows](https://github.com/monocongo/climate_indices/commit/3b040d59) (maintainer's own commit, same finding, landed ahead of this fix pass)

### Findings and fix status

#### [P2] EDDI's elementwise tail ran over the whole block — Fixed
- **Location:** `src/climate_indices/indices.py` (`eddi`, the mask/`np.clip`/`_hastings_inverse_normal`/final-clip sequence)
- **Origin:** Performance/memory lane
- **Problem:** the period mask, probability clip, Hastings inverse-normal approximation, and the `[-3.09, 3.09]` clip each allocated full-block temporaries. Measured on a 384 MB block (40 years monthly, 1e5 cells): peak RSS 4.47 GB, of which ~3.1 GB was the approximation alone (it allocates ~8 block-sized float64 arrays), i.e. peak ≈ 11.6× the input, with the block's width as the multiplier. Daily blocks are the worst case (366 periods). `percentage_of_normal` measured 2.13× on the same block, so this was EDDI-specific.
- **Applied fix:** the mask, clip, and approximation now run inside the existing per-period loop, one calendar period `(years, *cells)` at a time, and the final clip writes in place (`np.clip(..., out=eddi_values)`). The whole-block `np.where` mask and the `np.full` probability array are gone. Commit `9e25a76b`.

#### [P3] The rank-comparison budget did not bound wide grids — Fixed
- **Location:** `src/climate_indices/indices.py` (`_EDDI_RANK_COMPARISON_ELEMENT_BUDGET` and the rank loop)
- **Origin:** Correctness, performance, and dataflow lanes (same underlying claim, deduplicated)
- **Problem:** the comparison was chunked along climatology rows with a `max(1, budget // (years × cells))` floor, so once `years × cells` exceeded the budget a single chunk still allocated `years × cells` booleans — unbounded in the grid width, contradicting the constant's own comment. 40 years × 1e6 cells → 40 MB per chunk, ×40 climatology iterations, per period.
- **Applied fix:** landed ahead of this fix pass in the maintainer's `3b040d59`, which chunks across cells (`budget // (climatology years × years)`), so the bound holds for any grid width. This fix pass corrected the one doc sentence that still described the old axis (`benchmarks/README.md`) and re-verified the loop's equivalence with a one-cell chunk (`_EDDI_RANK_COMPARISON_ELEMENT_BUDGET` monkeypatched to 17).

#### [P3] `benchmarks/README.md` named the wrong chunk axis — Fixed
- **Location:** `benchmarks/README.md` (#942 conversion-status section)
- **Origin:** Maintainability lane
- **Problem:** after `3b040d59` the doc still said "climatology-row chunked count", so a reader tracing the mechanism from the doc found a different axis in the code.
- **Applied fix:** "cell-chunked count". Commit `9e25a76b`.

#### [P3] Restored the ambiguous-shape exception for spi/spei/pet — Fixed
- **Location:** `docs/xarray_compatibility.md` (spatial-kernel row)
- **Origin:** Maintainability lane
- **Problem:** the row rewrite dropped "rejects the one gridded shape that is ambiguous with it (a first cell axis of 12 or 366)" and replaced it with an unconditional "a 3-D or higher input is a block for spi/spei/pet", which `compute._prepare_input_shape` does not do for an undeclared ambiguous shape.
- **Applied fix:** the exception is back, alongside the new rule that EDDI and percentage of normal require the declaration for every 3-D input. Commit `9e25a76b`.

#### [P3] `percentage_of_normal` return layout — Fixed
- **Location:** `src/climate_indices/indices.py` (`percentage_of_normal` docstring)
- **Origin:** Maintainability lane
- **Problem:** the new `:return:` text said the output is "in the input's layout", but a 2-D input is still flattened to 1-D and returned as 1-D.
- **Applied fix:** the return text now names 1-D for 1-D/2-D input and `(time, *cells)` for a declared block. Commit `9e25a76b`.

#### [P4] Test coverage gaps — Fixed for the in-scope cases
- **Location:** `tests/test_spatial_kernel.py`
- **Origin:** Tests lane
- **Problem:** the new chunk loop always ran a single iteration on every fixture (≤6 cells), the EDDI no-loop guard was shape-only, and masked blocks, a 12-length first cell axis through the adapter, and the changed `DataShapeError.expected_shape` were unpinned.
- **Applied fix:** added a one-cell-chunk equivalence test (budget monkeypatched to 17), a masked-block test for each index (pinning the legacy "mask dropped" semantics, which differ from a missing value for EDDI), a grid whose first cell axis is 12 driven through the wired `typed_public_api` entry points (the adapter declares the reading; the NumPy API still refuses the shape), and an `expected_shape`/`actual_shape` assertion. The EDDI guard now pins one approximation call per calendar period `(years, 3, 2)` instead of one per block, matching the P2 fix. Commit `9e25a76b`.
- **Residual (not a defect):** the daily equivalence tests compare the spatial adapter against the per-cell adapter, so both sides share `_DailyCalendarPlan`; a calendar-plan bug would pass both. The monthly cases have an independent pointwise oracle. float32 block input is still unpinned (both paths promote identically, verified manually by the correctness lane).

#### [P2] Transposed dims + any NaN crash in the shared adapter — Blocked (pre-existing, needs its own ticket)
- **Location:** `src/climate_indices/xarray_adapter.py` (~lines 1835–1840, `sample_values = input_da.values; slices = [slice(None)] + [0] * (ndim - 1)`)
- **Origin:** Dataflow lane
- **Problem:** NaN validation samples the time series by assuming time is axis 0. For a `(lon, time, lat)` DataArray containing any NaN, the sample slices the `lon` axis and `_validate_calibration_non_nan_sample_size` raises `IndexError: boolean index did not match indexed array along axis 0; size of axis is 3 but size of corresponding boolean axis is 480`. Reproduced on this branch with `typed_public_api.spi` **and** `typed_public_api.eddi`, so it is not introduced by this PR — it affects every spatial kernel from #944/#951 onward and this PR's new capability as well. A NaN-free transposed grid works, since the validation is skipped.
- **Applied fix:** none in this PR. The fix (`input_da.isel({dim: 0 for dim in input_da.dims if dim != time_dim})`, or moving the core axis first before sampling) belongs in its own change against the adapter, which four other indices share — bundling it here would mix an adapter fix into a kernel-vectorization PR. Suggested follow-up: new issue referencing this finding, with the repro above.

### Remaining blockers
None for this PR. The one unresolved finding is the pre-existing adapter crash above, which is out of this PR's scope and needs a separate ticket; it does not block merging #977.

### Note on branch history
This branch also carries two commits from the maintainer's own review pass (`3b040d59` chunk-axis fix, `2afff302` complexity refactor of the EDDI/percentage-of-normal helpers), landed while this review was running. They are not part of this fix pass; the fix pass re-verified against the final HEAD (`9e25a76b`).

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/977
- **Lanes run:** 4 — correctness/numerical-equivalence, Python code quality/performance, test-coverage adequacy, architecture/API-contract consistency
- **Validation:** `uv run ruff check src/ tests/` clean; `uv run ruff format --check src/ tests/` (91 files formatted); `uv run mypy src/ tests/test_type_checking.py` no issues; `uv run pytest -q` — 2103 passed; `uv run pytest tests/test_release_integrity.py tests/test_review_scripts.py -q` — 45 passed; `uv run pytest tests/test_noaa_eddi_reference.py -m validation -q` (NCEI/NOAA oracle suite, excluded by default `addopts`) — 12 passed
- **Fix commits:**
  - [e5d4ece8 — fix(core): bound EDDI's per-period counts and suppress PNP's empty-slice warning](https://github.com/monocongo/climate_indices/commit/e5d4ece8335c6a0f775e5940901788665bae29a8)
  - [ed373b2f — docs(benchmarks): correct EDDI ranking-loop line citation](https://github.com/monocongo/climate_indices/commit/ed373b2f)
  - [f428ca58 — test(core): cover EDDI's all-missing block and a ragged chunk boundary](https://github.com/monocongo/climate_indices/commit/f428ca58)

### Findings and fix status

#### [P2] EDDI's per-period valid-/pad-counts weren't chunked, undercutting the PR's own memory-bound claim — Fixed
- **Location:** `src/climate_indices/indices.py` (EDDI's ranking setup, pre-fix ~line 435)
- **Origin:** Python quality/performance lane, independently re-verified
- **Problem:** The PR's stated design (comment + PR body: "the rank comparison is the kernel's memory ceiling... ~4 MB... rather than growing with the width of the spatial block") only chunked the inner `(climatology years, years, cells)` boolean comparison. `climatology_valid_counts`/`leading_pads_count` were precomputed once for *every* calendar period up front, sized `(climatology years, periods, *cells)` — for a wide daily grid (366 periods) this is ~366x larger than the chunked comparison it was meant to bound, so the stated "~4 MB regardless of grid width" guarantee didn't actually hold for the case the PR exists to serve (large gridded EDDI runs).
- **Applied fix:** Moved both counts inside the per-period loop so each is sized `(climatology years, *cells)` for one period at a time instead of all periods at once — removing the `num_periods` multiplier entirely rather than adding a second, separate chunking scheme. Commit `e5d4ece8`.

#### [P2] `percentage_of_normal` warns "Mean of empty slice" for any permanently-missing cell in an otherwise-valid block — Fixed
- **Location:** `src/climate_indices/indices.py`, `percentage_of_normal`'s calendar-period averaging (`np.nanmean` call, pre-fix ~line 1044)
- **Origin:** Test-coverage lane, reproduced independently
- **Problem:** The PR's own test suite treats "Mean of empty slice" as a real defect for a fully-missing *block* (asserted under `warnings.simplefilter("error")`), but the same warning fires for a single cell that's missing across the whole calibration period inside an otherwise-valid block — a realistic land/sea-mask pattern, and untested. Reproduced directly: a 3×2 grid with one all-NaN cell and the rest valid raises `RuntimeWarning: Mean of empty slice` under `-W error::RuntimeWarning`.
- **Applied fix:** Suppress that specific, expected warning around the averaging call; the resulting NaN for that cell is already correct and intentional. Commit `e5d4ece8`. Added regression test `test_permanently_missing_cell_does_not_warn`.

#### [P3] EDDI's `below` rank-count buffer reallocated and zero-filled every calendar period at a wider-than-needed dtype — Fixed
- **Location:** `src/climate_indices/indices.py`, EDDI ranking loop (pre-fix ~line 449)
- **Origin:** Python quality/performance lane
- **Problem:** `below = np.zeros(period_values.shape, dtype=np.int64)` was allocated fresh inside the `for period_index` loop (up to 366 times for a daily block) even though the following chunk loop always overwrites every element, and `int64` is twice as wide as the counts it holds ever require (bounded by the number of calibration years).
- **Applied fix:** Hoisted the buffer above the loop as `np.empty((num_years, cells_per_time_step), dtype=np.int32)`, reused across periods. Commit `e5d4ece8`.

#### [P4] Stale `indices.py` line-number citation in `benchmarks/README.md` — Fixed
- **Location:** `benchmarks/README.md:234`
- **Origin:** Architecture/API-contract lane
- **Problem:** This PR's docstring/logic changes to `eddi()` shifted the ranking loop off the line this doc cited (`indices.py:352`, which after the PR landed inside the new `spatial_time_major` docstring text, not the loop).
- **Applied fix:** Updated the citation to the loop's current line. Commits `ed373b2f` (initial correction) and `f428ca58`'s predecessor `e5d4ece8` (which shifted it again by restructuring the loop) made a second correction necessary; final value is `indices.py:446`, verified against the final HEAD after all fixes landed.

#### [P3] Test coverage: no test exercised EDDI's all-missing-block short-circuit, or a chunk-loop boundary that isn't a single cell — Fixed
- **Location:** `tests/test_spatial_kernel.py`
- **Origin:** Test-coverage lane
- **Problem:** `percentage_of_normal` had a dedicated `test_all_missing_block_short_circuits` test (with `warnings.simplefilter("error")`) but EDDI's identical short-circuit had no counterpart. Separately, the existing chunk-boundary test (`test_eddi_rank_comparison_chunks_across_cells`) forced `cells_per_chunk == 1`, the one case where a chunk can never be short relative to the grid — a ragged final chunk (grid width not a multiple of the chunk size) was untested.
- **Applied fix:** Added `test_all_missing_block_short_circuits` for EDDI, and `test_eddi_rank_comparison_handles_a_ragged_final_chunk` (budget chosen so a 6-cell grid chunks as 4 then 2). Commit `f428ca58`.

#### [P4] `_log_calculation_completed`'s docstring says it's "shared by every index function," but only `eddi()`/`percentage_of_normal()` call it — Not changed (no defect)
- **Location:** `src/climate_indices/indices.py:163` (docstring), duplicated logging blocks remain in `spi()`, `spei()`, `pet()`, `pci()`
- **Origin:** Python quality/performance lane
- **Problem:** The lane flagged this as worth a deliberate call, not a functional defect: the helper's docstring could be read as claiming universal adoption when 4 other functions still inline the same logging block.
- **Disposition:** Read literally, the docstring describes the *event* (`"calculation_completed"`, with the same duration/shape/memory fields) as shared across every index function — which is factually true regardless of which functions call this particular helper to emit it. Migrating `spi()`/`spei()`/`pet()`/`pci()` to the new helper would touch code this PR doesn't otherwise touch, which `AGENTS.md` explicitly asks agents not to do ("Scope new conventions to new code; do not migrate unrelated legacy code"). No change applied; flagged here for visibility rather than silently dropped.

#### [P3] No test compares the spatial-block EDDI path against the real NOAA/PSL oracle data — Not applied, follow-up recommended
- **Location:** `tests/test_noaa_eddi_reference.py`
- **Origin:** Test-coverage lane
- **Problem:** The oracle fixture (real NOAA-computed EDDI values, marked `FR-TEST-004`, "a critical test that must pass before merge") only ever exercises `indices.eddi()` on 1-D input. All new block-vs-something comparisons in this PR's test suite compare the rewritten code against itself (the per-cell adapter loop, or the same kernel called once per 1-D cell), which would not catch a bug confined to the block-specific reshape/pad/chunk plumbing if it happened to be self-consistent.
- **Disposition not applied because:** this repo treats scientific-validation fixtures as a distinct, more deliberate category (`VALIDATION.md`, referenced from `AGENTS.md`) rather than an ordinary unit-test surface, and the correctness lane independently produced a line-by-line mathematical proof that the block path's rank computation is algebraically identical to the oracle-validated 1-D path (same comparison, same NaN handling, same Tukey/Hastings formula, applied per-period regardless of block shape) — a form of verification at least as strong as a synthetic block built by duplicating the existing 1-D fixture across cells. Recommend the PR author or maintainer decide whether to extend the formal oracle fixture to a block case as a follow-up, rather than this review adding to a `FR`-numbered validation surface unilaterally.

### Remaining blockers
None. All P0–P3 findings that required a code or test change were fixed and verified (full suite, lint, type-check, and the NOAA oracle suite all pass on the final commit `f428ca58`); the two remaining items are a documented no-op (P4) and a scoped-out follow-up recommendation (P3), neither of which blocks merge.

File: src/climate_indices/indices.py [hunk 2]
```diff
@@ -919,6 +1013,7 @@ def percentage_of_normal(
             periodicity,
             clip_negatives=False,
             reshape=False,
+            spatial_time_major=spatial_time_major,
         )
 
         # extract the timesteps over which we'll compute the normal
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/indices.py:943]: Fixed in be511cee: `prepare_scaled` now converts a partially masked input to NaN right after shape prep, before `sum_to_scale` runs, so the spatial branch's `np.concatenate` no longer drops the mask and lets fill values leak into sliding sums, calibration normals, or percentages. Updated `test_eddi_masked_block_drops_the_mask` and `test_percentage_of_normal_masked_block_drops_the_mask` in `tests/test_spatial_kernel.py` to expect missing results for scale windows overlapping the mask, and verified against the per-cell path with the same mask.

File: src/climate_indices/indices.py [hunk 3]
```diff
@@ -932,19 +1027,35 @@ def percentage_of_normal(
...
+            # cell that is missing for the whole calibration period (not just the whole
+            # block, which is short-circuited above) is an expected all-NaN slice, not
+            # an error, so its "Mean of empty slice" warning is suppressed rather than
+            # left to propagate under a warnings-as-errors configuration.
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", message="Mean of empty slice", category=RuntimeWarning)
+                averages = np.nanmean(
+                    calibration_period_sums.reshape(-1, period_length, *calibration_period_sums.shape[1:]),
+                    axis=0,
+                )
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/indices.py]: Fixed in db3be9d0: replaced the `warnings.catch_warnings()`/`nanmean` block with an explicit count/sum reduction (`np.nansum` divided by `np.maximum(valid_counts, 1)`, NaN where `valid_counts == 0`), so an all-NaN calibration slice no longer needs a process-global warning filter that concurrent Dask blocks could race. `import warnings` was also removed from `indices.py` since it's now unused there.

File: src/climate_indices/indices.py [hunk 4]
```diff
@@ -759,8 +845,15 @@ def percentage_of_normal(
...
+    :param spatial_time_major: read a three-or-more-dimensional ``values`` as a
+        time-major block of independent time series, shaped (time, *cells), and
+        divide every cell by its own calendar-period normals in one pass. The
+        xarray adapter sets this for every block it packs; a direct NumPy caller
+        has to declare it, since a block is not a shape this function reads
+        without being told.
     :return: percent of normal precipitation values corresponding to the
-        scaled precipitation values array
+        scaled precipitation values array: 1-D for a 1-D or 2-D input, or the
+        (time, *cells) layout of a declared block
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/indices.py]: Fixed in 81fe7b60: converted the `percentage_of_normal` docstring from reST `:param:`/`:return:`/`:rtype:` fields to Google-style `Args:`/`Returns:` sections, including `spatial_time_major`.

**Comment by monocongo** [comment]
Addressed the outside-diff finding from [this review](https://github.com/monocongo/climate_indices/pull/977#pullrequestreview-5237717059) on `benchmarks/README.md`.

**Line ~235** ("core kernels take one 1-D temporal series... support for a spatial/cell axis comes only from the dispatch sites above") was stale even before this PR — `indices.spi`/`indices.spei` and the PET kernels already accept a declared `(time, *cells)` block directly via their own `spatial_time_major` parameter, and this PR adds the same to `indices.eddi`/`indices.percentage_of_normal`. Fixed in 81fe7b60 to name all five kernels that take a block directly.

**Line 138** is left as-is: that table is an explicitly frozen audit "as of `02c9cb38`" (see the intro at line 126-129), and the `### Conversion status (#923)` section a few paragraphs below is the living record of what's since been converted — it already documents `spi`/`spei` (#923), PET (#941), and now `eddi`/`percentage_of_normal` (#942, added in this PR). Editing the frozen row would break that established convention (the same row was left untouched when #923 and #941 converted spi/spei and PET), and would duplicate what the Conversion status section already states correctly.
